File: lab3/task1/C/parrot.c
#include <stdlib.h>

typedef char const *(*PTRFUN)();

struct Parrot
{
    PTRFUN* vtable;
    char const* name;
};

char const* name(void* this)
{
    return ((struct Parrot*) this)->name;
}

char const* greet()
{
    return "Sto mu gromova!";
}

char const* menu()
{
    return "brazilske orahe";
}

PTRFUN ParrotVTable[3] = {
    (PTRFUN) name,
    (PTRFUN) greet,
    (PTRFUN) menu
};

void* create(char const* name)
{
    struct Parrot* parrot = malloc( sizeof(struct Parrot) );
    parrot->name = name;
    parrot->vtable = ParrotVTable;
    return parrot;
}

File: lab3/task1/C#/Plugins/Parrot.cs
namespace MyFactory.Plugins
{
    public class Parrot : Animal
    {
        public override string Greeting { get ; set; } = "Sto mu gromova!";
        public override string Menu { get; set; } = "brazilske orahe";

        public Parrot(string name) : base(name) {}
    }
}